2p revamp (#1785)
cpinitiative/usaco-guide: content/3_Silver/2P.mdx

	</Resource>
</Resources>

## Two Pointers

The two pointers method iterates two pointers across an array, to track the start and end
of an interval. It can also be used to track two values in an array as shown in [CPH's 2SUM](https://usaco.guide/CPH.pdf#page=87) solution. 

<FocusProblem problem="book" />

## Solution - Books

We want to find the longest contiguous segment of books that can be read within $t$ minutes. 

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.

For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.

We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.

We can then continually increment $\texttt{left}$ by 1.

> **The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**

Reviewer: I would have said something like "At each timestep, exactly one pointer is incremented. Since each pointer can be incremented at most N times, it takes at most 2N time, resulting in a time complexity of $\mathcal{O}(N)$.
```suggestion
**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
```

Reviewer: but what you said is fine as well

	</Resource>
</Resources>

<FocusProblem problem="book" />

### Solution - Books

We want to find the longest continuous segment of books that can be read within $t$ minutes. 

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.

> 

Reviewer 2: Think it would help to add another line saying Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, binary search)

	</Resource>
</Resources>

## Two Pointers

The two pointers method iterates two pointers across an array, to track the start and end
of an interval, or two values in a sorted array that we are currently checking. Both pointers

> are monotonic; meaning each pointer starts at one end of the array and only moves in one

Reviewer 3: don't think monotonic is the right word here

<FocusProblem problem="book" />

## Solution - Books

We want to find the longest continuous segment of books that can be read within $t$ minutes. 

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.

> Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).

Reviewer 4: ok maybe it's just me, but the binary search part might distract from the focus of the problem.

Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, [binary search](/silver/binary-search)).

For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.

We will keep track of the total time with $\texttt{cur}$.

> Once $\texttt{cur}$ exceeds $t$, we remove the current $\texttt{right}$-indexed book and decrement $\texttt{right}$ by 1.

Reviewer 3: seems kinda extra wordy in the last two sentences. Both lines are kinda redundant so just remove both sentences.

> We will keep track of the max segment size with $\texttt{ans}$, and take the maximum between $\texttt{ans}$ and $\texttt{right} - \texttt{left}$.

Reviewer 3: also kind of self-explanatory that we are trying to find max.

Reviewer 3: maybe it's better to reiterate pointers move in one direction and the right pointer simply keeps moving to the right instead of starting over?

> You can visualize these pointers as maintaing a sliding window of books for this problem.

Reviewer 4: ```suggestion
You can visualize these pointers as maintaining a sliding window of books for this problem.
```

</Info>

### Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;

using ll = long long;

using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()

using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair

void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}

int main() {

> 	setIO();

Reviewer 4: since this is the focus problem, I think it would be beneficial to comment the code.

	</Resource>
</Resources>

## Two Pointers

The two pointers method iterates two pointers across an array, to track the start and end
of an interval. It can also be used to track two values in an array as shown in [CPH's 2SUM](https://usaco.guide/CPH.pdf#page=87) solution. 

<FocusProblem problem="book" />

## Solution - Books

We want to find the longest contiguous segment of books that can be read within $t$ minutes. 

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.

For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.

We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.

We can then continually increment $\texttt{left}$ by 1.

**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**

> <Info title="Pro-tip">

Reviewer 5: ```suggestion
<Info title="Pro Tip">
```
to be consistent

You can visualize these pointers as maintaining a sliding window of books for this problem.

</Info>

### Implementation

<LanguageSection>

<CPPSection>

> 

Reviewer 5: ```suggestion
```

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;

using ll = long long;

using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()

using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair

> void setIO(string name = "") {

Reviewer 5: there's a codesnip for all this; wonder why we never use it?? it's `CPP Short Template`

We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$ and increment $\texttt{left}$ by 1.

**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**

<Info title="Pro-tip">

You can visualize these pointers as maintaing a sliding window of books for this problem.

</Info>

### Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;

using ll = long long;

using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()

using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair

void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}

int main() {
	setIO();
	int n, t, ans = 0;
	cin >> n >> t;
	vi books(n);
	for(int i = 0; i < n; i++){
		cin >> books[i];
	}

	int left = 0, right = 0, cur = 0;
	while(left < n && right < n) {

> 		while(right < n) {

Reviewer 6: oh didn't see these two while loops, could you add spaces for these

			cur += books[right++];
			if (cur > t) {
				cur -= books[--right]; 
				break;
			}
		}

		ans = max(ans, right - left);
		cur -= books[left++];
	}

	cout << ans;
}
```

</CPPSection>

<JavaSection>

```java
import java.util.*;
import java.io.*;

public class Books {
	public static void main(String[] args) {
		Kattio io = new Kattio();

		int n = io.nextInt();
		int t = io.nextInt();
		int[] books = new int[n];

> 		for(int i = 0; i < n; i++) {

Reviewer 6: and this for loop lmao
```suggestion
		for (int i = 0; i < n; i++) {
```

	</Resource>
</Resources>

## Two Pointers

The two pointers method iterates two pointers across an array, to track the start and end

> of an interval. It can also be used to track two values in order to satisfy a specific condition. 

Reviewer 3: still sorta unclear. To a new reader with no knowledge of 2p, "used to trace two values in order to satisfy a specific condition" is simply too broad and isn't very helpful.

> of an interval. It can also be used to track two values in an array as show in CPH's 2SUM solution. 

Reviewer 5: ```suggestion
of an interval. It can also be used to track two values in an array as shown in CPH's 2SUM solution. 
```

> of an interval. It can also be used to track two values in an array as shown in CPH's 2SUM solution. 

Reviewer 3: maybe just link cph again?

<FocusProblem problem="book" />

## Solution - Books

> We want to find the longest continuous segment of books that can be read within $t$ minutes. 

Reviewer 5: is contiguous a better word? idk

Author: sure.

> We want to find the longest continuous segment of books that can be read within $t$ minutes. 

Author: ```suggestion
We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
```

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.

Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).

> For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.

Reviewer 3: > For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
Kind of misleading - you're not really increasing unit it exceeds. It's more like increasing until it reaches max.

Author: wdym by the max? im increasing it until it exceeds t right?

Reviewer 3: well it isn't really _exceeding_ t right - it's more like increasing until it hits the maximum that's less than t.

Author: > well it isn't really _exceeding_ t right - it's more like increasing until it hits the maximum that's less than t.

@andrewwangva oh i c what u mean, but for each time step, we are increasing right until it exceeds t tho.

Author: oh wait i misread ur msg, ill make it clearer

Author: okay fixed.

We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.

> We can then increment $\texttt{left}$ by 1.

Reviewer 3: maybe say continually increment one to clarify?

> We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.

Reviewer 3: Hard to follow. Instead focus on the continuous movement of the left pointer and at each moment the calculation of max books read.

> We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.

> **The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**

Reviewer 7: > **The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**

add that because of this, the time complexity is O(N)

Reviewer 7: not addressed? also, it might be better to say that the time complexity is O(N) because the left and right pointers move at most N steps in total.

Author: i moved the time comp down cus i thought i was a little more consistent (?)

Author: ill just move it up in the explanation ig

	</Resource>
</Resources>

## Two Pointers

The two pointers method iterates two pointers across an array, to track the start and end
of an interval, or two values in a sorted array that we are currently checking. 
Typically, each pointer starts at seperate ends and moves strictly in one direction. 

> However, as shown in the focus problem below, the two pointers do not have to start at seperate ends.

Reviewer 8: oops I found this rather confusing (no idea how to fix though) D:

Author: which part is confusing?

<FocusProblem problem="book" />

## Solution - Books

We want to find the longest continuous segment of books that can be read within $t$ minutes. 

To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.

> Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, [binary search](/silver/binary-search)).

Reviewer 8: ```suggestion
Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
```

For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.

We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.

> We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$ and increment $\texttt{left}$ by 1.

Reviewer 8: we never defined cur. maybe write this out explicitly?

</Info>

### Implementation

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;

using ll = long long;

using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()

using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair

void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}

int main() {
	setIO();
	int n, t, ans = 0;
	cin >> n >> t;
	vi books(n);
	for(int i = 0; i < n; i++){
		cin >> books[i];
	}

	int left = 0, right = 0, cur = 0;
	while(left < n && right < n) {
		while(right < n) {
			cur += books[right++];
			if(cur > t) {
				cur -= books[--right]; 
				break;
			}
		}

		ans = max(ans, right - left);
		cur -= books[left++];
	}

	cout << ans;
}
```

</CPPSection>

<JavaSection>

```java
import java.util.*;
import java.io.*;

public class Books {
	public static void main(String[] args) {
		Kattio io = new Kattio();

> 		int n = io.nextInt();
> 		int t = io.nextInt();
> 
> 		int[] books = new int[n];
> 
> 		for(int i = 0; i < n; i++) {
> 			books[i] = io.nextInt();
> 		}

Author: ```suggestion
		int n = io.nextInt();
		int t = io.nextInt();
		int[] books = new int[n];
		for(int i = 0; i < n; i++) {
			books[i] = io.nextInt();
		}
```

		int left = 0, right = 0, cur = 0, ans = 0;

> 		while(left < n && right < n) {

Reviewer 6: ```suggestion
		while (left < n && right < n) {
```

> 			while(right < n) {

Reviewer 6: ```suggestion
			while (right < n) {
```

				cur += books[right++];

> 				if(cur > t) {

Reviewer 6: if (cur > t) {

					cur -= books[--right]; 
					break;
				}
			}

			ans = Math.max(ans, right - left);
			cur -= books[left++];
		}

		io.println(ans);

> 		io.close();

Reviewer 6: 139 and 141 seem unnecessary
